man/asciidoctor.1
'\" t
.\"     Title: asciidoctor
.\"    Author: Dan Allen, Sarah White
.\" Generator: Asciidoctor 2.0.25
.\"      Date: 2018-03-20
.\"    Manual: Asciidoctor Manual
.\"    Source: Asciidoctor 2.0.26
.\"  Language: English
.\"
.TH "ASCIIDOCTOR" "1" "2018-03-20" "Asciidoctor 2.0.26" "Asciidoctor Manual"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
asciidoctor \- converts AsciiDoc source files to HTML, DocBook, and other formats
.SH "SYNOPSIS"
.sp
\fBasciidoctor\fP [\fIOPTION\fP].\|.\|. \fIFILE\fP.\|.\|.
.SH "DESCRIPTION"
.sp
The asciidoctor(1) command converts the AsciiDoc source file(s) \fIFILE\fP to HTML5, DocBook 5, man(ual) page, and other custom output formats.
.sp
If \fIFILE\fP is \fI\-\fP then the AsciiDoc source is read from standard input.
.SH "OPTIONS"
.SS "Security Settings"
.sp
\fB\-B, \-\-base\-dir\fP=\fIDIR\fP
.RS 4
Base directory containing the document and resources.
Defaults to the directory containing the source file or, if the source is read from a stream, the working directory.
When combined with the safe mode setting, can be used to chroot the execution of the program.
.RE
.sp
\fB\-S, \-\-safe\-mode\fP=\fISAFE_MODE\fP
.RS 4
Set safe mode level: \fIunsafe\fP, \fIsafe\fP, \fIserver\fP, or \fIsecure\fP.
Disables potentially dangerous macros in source files, such as \f(CRinclude::[]\fP.
If not set, the safe mode level defaults to \fIunsafe\fP when Asciidoctor is invoked using this script.
.RE
.sp
\fB\-\-safe\fP
.RS 4
Set safe mode level to \fIsafe\fP.
Enables include directives, but prevents access to ancestor paths of source file.
Provided for compatibility with the asciidoc command.
If not set, the safe mode level defaults to \fIunsafe\fP when Asciidoctor is invoked using this script.
.RE
.SS "Document Settings"
.sp
\fB\-a, \-\-attribute\fP=\fIATTRIBUTE\fP
.RS 4
Define, override, or unset a document attribute.
Command\-line attributes take precedence over attributes defined in the source file unless either the name or value ends in \fI@\fP.
No substitutions are applied to the value.
.sp
\fIATTRIBUTE\fP is normally formatted as a key\-value pair, in the form \fINAME=VALUE\fP.
Alternate forms are \fINAME\fP (where the \fIVALUE\fP defaults to an empty string), \fINAME!\fP (unsets the \fINAME\fP attribute), and \fINAME=VALUE@\fP (or \fINAME@=VALUE\fP) (where \fIVALUE\fP does not override the \fINAME\fP attribute if it\(cqs already defined in the source document).
A value containing spaces must be enclosed in quotes, in the form \fINAME="VALUE WITH SPACES"\fP.
.sp
This option may be specified more than once.
.RE
.sp
\fB\-b, \-\-backend\fP=\fIBACKEND\fP
.RS 4
Backend output file format: \fIhtml5\fP, \fIdocbook5\fP, and \fImanpage\fP are supported out of the box.
You can also use the backend alias names \fIhtml\fP (aliased to \fIhtml5\fP) or \fIdocbook\fP (aliased to \fIdocbook5\fP).
Other values can be passed, but if Asciidoctor cannot resolve the backend to a converter, it will fail.
Defaults to \fIhtml5\fP.
.RE
.sp
\fB\-d, \-\-doctype\fP=\fIDOCTYPE\fP
.RS 4
Document type: \fIarticle\fP, \fIbook\fP, \fImanpage\fP, or \fIinline\fP.
Sets the root element when using the \fIdocbook\fP backend and the style class on the HTML body element when using the \fIhtml\fP backend.
The \fIbook\fP document type allows multiple level\-0 section titles in a single document.
The \fImanpage\fP document type enables parsing of metadata necessary to produce a man page.
The \fIinline\fP document type allows the content of a single paragraph to be formatted and returned without wrapping it in a containing element.
Defaults to \fIarticle\fP.
.RE
.SS "Document Conversion"
.sp
\fB\-D, \-\-destination\-dir\fP=\fIDIR\fP
.RS 4
Destination output directory.
Defaults to the directory containing the source file or, if the source is read from a stream, the working directory.
If specified, the directory is resolved relative to the working directory.
.RE
.sp
\fB\-E, \-\-template\-engine\fP=\fINAME\fP
.RS 4
Template engine to use for the custom converter templates.
The gem with the same name as the engine will be loaded automatically.
This name is also used to build the full path to the custom converter templates.
If a template engine is not specified, it will be auto\-detected based on the file extension of the custom converter templates found.
.RE
.sp
\fB\-e, \-\-embedded\fP
.RS 4
Output an embeddable document, which excludes the header, the footer, and everything outside the body of the document.
This option is useful for producing documents that can be inserted into an external template.
.RE
.sp
\fB\-I, \-\-load\-path\fP=\fIDIRECTORY\fP
.RS 4
Add the specified directory to the load path, so that \fI\-r\fP can load extensions from outside the default Ruby load path.
This option may be specified more than once.
.RE
.sp
\fB\-n, \-\-section\-numbers\fP
.RS 4
Auto\-number section titles.
Synonym for \fB\-\-attribute sectnums\fP.
.RE
.sp
\fB\-o, \-\-out\-file\fP=\fIOUT_FILE\fP
.RS 4
Write output to file \fIOUT_FILE\fP.
Defaults to the base name of the input file suffixed with \fIbackend\fP extension.
The file is resolved relative to the working directory.
If the input is read from standard input or a named pipe (fifo), then the output file defaults to stdout.
If \fIOUT_FILE\fP is \fI\-\fP, then the output file is written to standard output.
.RE
.sp
\fB\-R, \-\-source\-dir\fP=\fIDIR\fP
.RS 4
Source directory.
Currently only used if the destination directory is also specified.
Used to preserve the directory structure of files converted within this directory in the destination directory.
If specified, the directory is resolved relative to the working directory.
.RE
.sp
\fB\-r, \-\-require\fP=\fILIBRARY\fP
.RS 4
Require the specified library before executing the processor, using the standard Ruby require.
This option may be specified more than once.
.RE
.sp
\fB\-s, \-\-no\-header\-footer\fP
.RS 4
Output an embeddable document, which excludes the header, the footer, and everything outside the body of the document.
This option is useful for producing documents that can be inserted into an external template.
.RE
.sp
\fB\-T, \-\-template\-dir\fP=\fIDIR\fP
.RS 4
A directory containing custom converter templates that override one or more templates from the built\-in set.
(requires \fItilt\fP gem)
.sp
If there is a subfolder that matches the engine name (if specified), that folder is appended to the template directory path.
Similarly, if there is a subfolder in the resulting template directory that matches the name of the backend, that folder is appended to the template directory path.
.sp
This option may be specified more than once.
Matching templates found in subsequent directories override ones previously discovered.
.RE
.SS "Processing Information"
.sp
\fB\-\-failure\-level\fP=\fILEVEL\fP
.RS 4

CHANGELOG.adoc
= Asciidoctor Changelog
:url-asciidoctor: https://asciidoctor.org
:url-asciidoc: https://docs.asciidoctor.org/asciidoc/latest/
:url-repo: https://github.com/asciidoctor/asciidoctor
:icons: font
:star: icon:star[role=red]
ifndef::icons[]
:star: &#9733;
endif::[]

{url-asciidoctor}[Asciidoctor] is a _fast_, open source text processor and publishing toolchain for converting {url-asciidoc}[AsciiDoc] content into HTML 5, DocBook 5, and other formats.

This document provides a high-level view of the changes introduced in Asciidoctor by release.
For an even more detailed look at what has changed, refer to the {url-repo}/commits/[commit history] on GitHub.

This project utilizes semantic versioning.

// tag::compact[]
== Unreleased

Enhancements::

  * Introduce 4+ tilde (`~`) delimiter for open blocks; restrict masquerading to partintro and abstract (#1121)
  * Allow the CLI to control the logging level using the `--log-level` option (#3868)
  * Ignore `--log-level` CLI option if `-q` is specified; ignore `-v` CLI option if `--log-level` is specified (#3868)
  * Add support for inline linenums mode when using Rouge as source highlighter (`rouge-linenums-mode=inline`) (#3641)
  * Look up asset dir attribute (i.e., `imagesdir` or `iconsdir`) on node to resolve asset path prefix, inheriting from document if not otherwise set (#3661) (*@mogztter*)
  * Omit meta generator tag in HTML output if `reproducible` document attribute is set (#4143) (*@rpavlik*)
  * Add `--sourcemap` option to the CLI to enable the `:sourcemap` option on the parser (#3569) (*@hackingotter*)
  * Allow section ID to be unset by assigning an empty value to the `id` block attribute (#4139)
  * Allow section to override toclevels for itself and descendant sections (#2618)
  * Add support for dropping principal text of list item if empty in an ordered or unordered list when converting to man page (#4182)
  * Don't add empty line before block content if principal text of dlist item is not specified when converting to man page (#4182)
  * Add support for role on thematic break in HTML converter (#4101)
  * Add support for link=self on image macros (#3656)
  * Alias Inline#content to Inline#text so Inline node quacks like other nodes (#3220)
  * Update signature of Logger constructor so both positional arguments and keyword arguments are passed to super constructor (#4250)
  * Honor `level` and `formatter` keyword arguments passed to Logger constructor (#4250)
  * Add support for skipping TOML front matter (Hugo) when `skip-front-matter` attribute is set (#4300) (*@abhinav*)
  * Add support for Wistia using the video macro
  * Skip front matter in included content if skip-front-matter option is set on include directive (#3437)
  * Allow handles? method on include processor to accept doc as first argument; mixin adapter to support old method signature (#2895)
  * Take start value into account when validating explicit numbering on ordered list (#3252)
  * Allow ordered list with explicit numbering to set start value using first item (#2218)
  * Pass through id attribute defined on inline image to output (#4311)
  * Add cxx attribute for C++ as preferred alternative to cpp (#4442)
  * Don't add highlight class to pre tag when `nohighlight` option is set (highlight.js only) (#4804)

Compliance::

  * Don't promote level-0 special section at start of document to document title (#4151)
  * Disallow the use of dot (`.`) in the name of a named element attribute (#4147)

lib/asciidoctor.rb
# frozen_string_literal: true

require 'set'

# NOTE RUBY_ENGINE == 'opal' conditional blocks like this are filtered by the Opal preprocessor
if RUBY_ENGINE == 'opal'
  # this require is satisfied by the Asciidoctor.js build; it augments the Ruby environment for Asciidoctor.js
  require 'asciidoctor/js'
else
  require 'cgi/util'
  autoload :OpenURI, 'open-uri'
  autoload :Pathname, 'pathname'
  autoload :StringScanner, 'strscan'
  autoload :URI, 'uri'
end

# Public: The main application interface (API) for Asciidoctor. This API provides methods to parse AsciiDoc content and
# convert it to various output formats using built-in or third-party converters or Tilt-supported templates.
#
# An AsciiDoc document can be as simple as a single line of content, though it more commonly starts with a document
# header that declares the document title and document attribute definitions. The document header is then followed by
# zero or more section titles, optionally nested, to organize the paragraphs, blocks, lists, etc. of the document.
#
# By default, the processor converts the AsciiDoc document to HTML 5 using a built-in converter. However, this behavior
# can be changed by specifying a different backend (e.g., +docbook+). A backend is a keyword for an output format (e.g.,
# DocBook). That keyword, in turn, is used to select a converter, which carries out the request to convert the document
# to that format.
#
# In addition to this API, Asciidoctor also provides a command-line interface (CLI) named +asciidoctor+ for converting
# AsciiDoc content. See the provided man(ual) page for usage and options.
#
# Examples
#
#   # Convert an AsciiDoc file
#   Asciidoctor.convert_file 'document.adoc', safe: :safe
#
#   # Convert an AsciiDoc string
#   puts Asciidoctor.convert "I'm using *Asciidoctor* version {asciidoctor-version}.", safe: :safe
#
#   # Convert an AsciiDoc file using Tilt-supported templates
#   Asciidoctor.convert_file 'document.adoc', safe: :safe, template_dirs: ['/path/to/templates']
#
#   # Parse an AsciiDoc file into a document object
#   doc = Asciidoctor.load_file 'document.adoc', safe: :safe
#
#   # Parse an AsciiDoc string into a document object
#   doc = Asciidoctor.load "= Document Title\n\nfirst paragraph\n\nsecond paragraph", safe: :safe
#
module Asciidoctor
  # alias the RUBY_ENGINE constant inside the Asciidoctor namespace and define a precomputed alias for runtime
  RUBY_ENGINE_OPAL = (RUBY_ENGINE = ::RUBY_ENGINE) == 'opal'

  module SafeMode
    # A safe mode level that disables any of the security features enforced
    # by Asciidoctor (Ruby is still subject to its own restrictions).
    UNSAFE = 0

    # A safe mode level that closely parallels safe mode in AsciiDoc. This value
    # prevents access to files which reside outside of the parent directory of
    # the source file and disables any macro other than the include::[] directive.
    SAFE = 1

    # A safe mode level that disallows the document from setting attributes
    # that would affect the conversion of the document, in addition to all the
    # security features of SafeMode::SAFE. For instance, this level forbids
    # changing the backend or source-highlighter using an attribute defined
    # in the source document header. This is the most fundamental level of
    # security for server deployments (hence the name).
    SERVER = 10

    # A safe mode level that disallows the document from attempting to read
    # files from the file system and including the contents of them into the
    # document, in additional to all the security features of SafeMode::SERVER.
    # For instance, this level disallows use of the include::[] directive and the
    # embedding of binary content (data uri), stylesheets and JavaScripts
    # referenced by the document. (Asciidoctor and trusted extensions may still
    # be allowed to embed trusted content into the document).
    #
    # Since Asciidoctor is aiming for wide adoption, this level is the default
    # and is recommended for server deployments.
    SECURE = 20

    # A planned safe mode level that disallows the use of passthrough macros and
    # prevents the document from setting any known attributes, in addition to all
    # the security features of SafeMode::SECURE.
    #
    # Please note that this level is not currently implemented (and therefore not
    # enforced)!
    #PARANOID = 100

    @names_by_value = (constants false).map {|sym| [(const_get sym), sym.to_s.downcase] }.sort {|(a), (b)| a <=> b }.to_h

    def self.value_for_name name
      const_get name.upcase, false rescue nil
    end

    def self.name_for_value value
      @names_by_value[value]
    end

    def self.names
      @names_by_value.values
    end
  end

  # Flags to control compliance with the behavior of AsciiDoc
  module Compliance
    @keys = ::Set.new
    class << self
      attr_reader :keys

      # Defines a new compliance key and assigns an initial value.
      def define key, value
        instance_variable_set %(@#{key}), value
        singleton_class.send :attr_accessor, key
        @keys << key
        nil
      end
    end

    # AsciiDoc terminates paragraphs adjacent to
    # block content (delimiter or block attribute list)
    # This option allows this behavior to be modified
    # TODO what about literal paragraph?
    # Compliance value: true
    define :block_terminates_paragraph, true

    # AsciiDoc does not parse paragraphs with a verbatim style
    # (i.e., literal, listing, source, verse) as verbatim content.
    # This options allows this behavior to be modified
    # Compliance value: false
    define :strict_verbatim_paragraphs, true

    # AsciiDoc supports both atx (single-line) and setext (underlined) section titles.
    # This option can be used to disable the setext variant.
    # Compliance value: true
    define :underline_style_section_titles, true

    # Asciidoctor will unwrap the content in a preamble if the document has a
    # title and no sections, then discard the empty preamble.
    # Compliance value: false
    define :unwrap_standalone_preamble, true

    # AsciiDoc drops lines that contain references to missing attributes.
    # This behavior is not intuitive to most writers.
    # Asciidoctor allows this behavior to be configured.
    # Possible options are 'skip', 'drop', 'drop-line', and 'warn'.
    # Compliance value: 'drop-line'
    define :attribute_missing, 'skip'

    # AsciiDoc drops lines that contain an attribute unassignment.
    # This behavior may need to be tuned depending on the circumstances.
    # Compliance value: 'drop-line'
    define :attribute_undefined, 'drop-line'

    # Asciidoctor will allow the id, role and options to be set
    # on blocks using a shorthand syntax (e.g., #idname.rolename%optionname)
    # Compliance value: false
    define :shorthand_property_syntax, true

    # Asciidoctor will attempt to resolve the target of a cross reference by
    # matching its reference text (reftext or title) (e.g., <<Section Title>>)
    # Compliance value: false
    define :natural_xrefs, true

    # Asciidoctor will start counting at the following number
    # when creating a unique id when there is a conflict
    # Compliance value: 2
    define :unique_id_start_index, 2

    # Asciidoctor will recognize commonly-used Markdown syntax
    # to the degree it does not interfere with existing
    # AsciiDoc syntax and behavior.
    # Compliance value: false
    define :markdown_syntax, true
  end

  # The absolute root directory of the Asciidoctor RubyGem
  ROOT_DIR = ::File.dirname ::File.absolute_path __dir__ unless defined? ROOT_DIR

  # The absolute lib directory of the Asciidoctor RubyGem
  LIB_DIR = ::File.join ROOT_DIR, 'lib'

  # The absolute data directory of the Asciidoctor RubyGem
  DATA_DIR = ::File.join ROOT_DIR, 'data'

  # The user's home directory, as best we can determine it
  # IMPORTANT this rescue is required for running Asciidoctor on GitHub.com
  USER_HOME = ::Dir.home rescue (::ENV['HOME'] || ::Dir.pwd) # rubocop:disable Style/EnvHome

  # The newline character used for output; stored in constant table as an optimization
  LF = ?\n

  # The null character to use for splitting attribute values
  NULL = ?\0

  # String for matching tab character
  TAB = ?\t

  # Maximum integer value for "boundless" operations; equal to MAX_SAFE_INTEGER in JavaScript
  MAX_INT = 9007199254740991

  # Alias UTF_8 encoding for convenience / speed
  UTF_8 = ::Encoding::UTF_8

  # Byte arrays for UTF-* Byte Order Marks
  BOM_BYTES_UTF_8 = [0xef, 0xbb, 0xbf]
  BOM_BYTES_UTF_16LE = [0xff, 0xfe]
  BOM_BYTES_UTF_16BE = [0xfe, 0xff]

  # The mode to use when opening a file for reading
  FILE_READ_MODE = RUBY_ENGINE_OPAL ? 'r' : 'rb:UTF-8:UTF-8'

  # The mode to use when opening a URI for reading
  URI_READ_MODE = FILE_READ_MODE

  # The mode to use when opening a file for writing
  FILE_WRITE_MODE = RUBY_ENGINE_OPAL ? 'w' : 'wb:UTF-8'

  # The default document type
  # Can influence markup generated by the converters
  DEFAULT_DOCTYPE = 'article'

  # The backend determines the format of the converted output, default to html5
  DEFAULT_BACKEND = 'html5'

  DEFAULT_STYLESHEET_KEYS = ::Set['', 'DEFAULT']

  DEFAULT_STYLESHEET_NAME = 'asciidoctor.css'

  # Pointers to the preferred version for a given backend.
  BACKEND_ALIASES = {
    'html' => 'html5',
    'docbook' => 'docbook5',
  }

  # Default page widths for calculating absolute widths
  DEFAULT_PAGE_WIDTHS = { 'docbook' => 425 }

  # Default extensions for the respective base backends
  DEFAULT_EXTENSIONS = {
    'html' => '.html',
    'docbook' => '.xml',
    'pdf' => '.pdf',
    'epub' => '.epub',
    'manpage' => '.man',
    'asciidoc' => '.adoc',
  }

  # A map of file extensions that are recognized as AsciiDoc documents
  # TODO .txt should be deprecated
  ASCIIDOC_EXTENSIONS = {
    '.adoc' => true,
    '.asciidoc' => true,
    '.asc' => true,
    '.ad' => true,
    # TODO .txt should be deprecated
    '.txt' => true,
  }

  SETEXT_SECTION_LEVELS = {
    '=' => 0,
    '-' => 1,
    '~' => 2,
    '^' => 3,
    '+' => 4,
  }

  ADMONITION_STYLES = ::Set['NOTE', 'TIP', 'IMPORTANT', 'WARNING', 'CAUTION']

  ADMONITION_STYLE_HEADS = ::Set.new(ADMONITION_STYLES.map {|s| s.chr })

  PARAGRAPH_STYLES = ::Set['comment', 'example', 'literal', 'listing', 'normal', 'open', 'pass', 'quote', 'sidebar', 'source', 'verse', 'abstract', 'partintro']

  VERBATIM_STYLES = ::Set['literal', 'listing', 'source', 'verse']

  DELIMITED_BLOCKS = {
    '--' => [:open, ::Set['comment', 'example', 'literal', 'listing', 'pass', 'quote', 'sidebar', 'source', 'verse', 'admonition', 'abstract', 'partintro']],
    '----' => [:listing, ::Set['literal', 'source']],
    '....' => [:literal, ::Set['listing', 'source']],
    '====' => [:example, ::Set['admonition']],
    '****' => [:sidebar, ::Set.new],
    '____' => [:quote, ::Set['verse']],
    '++++' => [:pass, ::Set['stem', 'latexmath', 'asciimath']],
    '|===' => [:table, ::Set.new],
    ',===' => [:table, ::Set.new],
    ':===' => [:table, ::Set.new],
    '!===' => [:table, ::Set.new],
    '////' => [:comment, ::Set.new],
    '```' => [:fenced_code, ::Set.new],
  }

  DELIMITED_BLOCK_HEADS = {}.tap {|accum| DELIMITED_BLOCKS.each_key {|k| accum[k.slice 0, 2] = true } }
  DELIMITED_BLOCK_TAILS = {}.tap {|accum| DELIMITED_BLOCKS.each_key {|k| accum[k] = k[k.length - 1] if k.length == 4 } }

  # NOTE the 'figure' key as a string is historical and used by image blocks
  CAPTION_ATTRIBUTE_NAMES = { example: 'example-caption', 'figure' => 'figure-caption', listing: 'listing-caption', table: 'table-caption' }

  LAYOUT_BREAK_CHARS = {
    '\'' => :thematic_break,
    '<' => :page_break,
  }

  MARKDOWN_THEMATIC_BREAK_CHARS = {
    '-' => :thematic_break,
    '*' => :thematic_break,
    '_' => :thematic_break,
  }

  HYBRID_LAYOUT_BREAK_CHARS = LAYOUT_BREAK_CHARS.merge MARKDOWN_THEMATIC_BREAK_CHARS

  #LIST_CONTEXTS = [:ulist, :olist, :dlist, :colist]

  NESTABLE_LIST_CONTEXTS = [:ulist, :olist, :dlist]

  # TODO validate use of explicit style name above ordered list (this list is for selecting an implicit style)
  ORDERED_LIST_STYLES = [:arabic, :loweralpha, :lowerroman, :upperalpha, :upperroman] #, :lowergreek]

  ORDERED_LIST_KEYWORDS = {
    #'arabic' => '1',
    #'decimal' => '1',
    'loweralpha' => 'a',
    'lowerroman' => 'i',
    #'lowergreek' => 'a',
    'upperalpha' => 'A',
    'upperroman' => 'I',
  }

  ATTR_REF_HEAD = '{'

  LIST_CONTINUATION = '+'

  # NOTE AsciiDoc.py allows + to be preceded by TAB; Asciidoctor does not
  HARD_LINE_BREAK = ' +'

  LINE_CONTINUATION = ' \\'

  LINE_CONTINUATION_LEGACY = ' +'

  BLOCK_MATH_DELIMITERS = {
    asciimath: ['\$', '\$'],
    latexmath: ['\[', '\]'],
  }

  INLINE_MATH_DELIMITERS = {
    asciimath: ['\$', '\$'],
    latexmath: ['\(', '\)'],
  }

  (STEM_TYPE_ALIASES = {
    'latexmath' => 'latexmath',
    'latex' => 'latexmath',
    'tex' => 'latexmath',
  }).default = 'asciimath'

  FONT_AWESOME_VERSION = '4.7.0'

  HIGHLIGHT_JS_VERSION = '9.18.3'

  MATHJAX_VERSION = '2.7.9'

  DEFAULT_ATTRIBUTES = {
    'appendix-caption' => 'Appendix',
    'appendix-refsig' => 'Appendix',
    'caution-caption' => 'Caution',
    'chapter-refsig' => 'Chapter',
    #'encoding' => 'UTF-8',
    'example-caption' => 'Example',
    'figure-caption' => 'Figure',
    'important-caption' => 'Important',
    'last-update-label' => 'Last updated',
    #'listing-caption' => 'Listing',
    'note-caption' => 'Note',
    'part-refsig' => 'Part',
    #'preface-title' => 'Preface',
    'prewrap' => '',
    'sectids' => '',
    'section-refsig' => 'Section',
    'table-caption' => 'Table',
    'tip-caption' => 'Tip',
    'toc-placement' => 'auto',
    'toc-title' => 'Table of Contents',
    'untitled-label' => 'Untitled',
    'version-label' => 'Version',
    'warning-caption' => 'Warning',
  }

  # attributes which be changed throughout the flow of the document (e.g., sectnums)
  FLEXIBLE_ATTRIBUTES = ['sectnums']

  INTRINSIC_ATTRIBUTES = {
    'startsb' => '[',
    'endsb' => ']',
    'vbar' => '|',
    'caret' => '^',
    'asterisk' => '*',
    'tilde' => '~',
    'plus' => '&#43;',
    'backslash' => '\\',
    'backtick' => '`',
    'blank' => '',
    'empty' => '',
    'sp' => ' ',
    'two-colons' => '::',
    'two-semicolons' => ';;',
    'nbsp' => '&#160;',
    'deg' => '&#176;',
    'zwsp' => '&#8203;',
    'quot' => '&#34;',
    'apos' => '&#39;',
    'lsquo' => '&#8216;',
    'rsquo' => '&#8217;',
    'ldquo' => '&#8220;',
    'rdquo' => '&#8221;',
    'wj' => '&#8288;',
    'brvbar' => '&#166;',
    'pp' => '&#43;&#43;',
    'cpp' => 'C&#43;&#43;',
    'cxx' => 'C&#43;&#43;',
    'amp' => '&',
    'lt' => '<',
    'gt' => '>',
  }

  # Regular expression character classes (to ensure regexp compatibility between Ruby and JavaScript)
  # CC stands for "character class", CG stands for "character class group"
  unless RUBY_ENGINE == 'opal'
    # CC_ALL is any character, including newlines (must be accompanied by multiline regexp flag)
    CC_ALL = '.'
    # CC_ANY is any character except newlines
    CC_ANY = '.'
    CC_EOL = '$'
    CC_ALPHA = CG_ALPHA = '\p{Alpha}'
    CC_ALNUM = CG_ALNUM = '\p{Alnum}'
    CG_BLANK = '\p{Blank}'
    CC_WORD = CG_WORD = '\p{Word}'
  end

  QuoteAttributeListRxt = %(\\[([^\\[\\]]+)\\])

  QUOTE_SUBS = {}.tap do |accum|
    # unconstrained quotes:: can appear anywhere
    # constrained quotes:: must be bordered by non-word characters
    # NOTE these substitutions are processed in the order they appear here and
    # the order in which they are replaced is important
    accum[false] = normal = [
      # **strong**
      [:strong, :unconstrained, /\\?(?:#{QuoteAttributeListRxt})?\*\*(#{CC_ALL}+?)\*\*/m],
      # *strong*
      [:strong, :constrained, /(^|[^#{CC_WORD};:}])(?:#{QuoteAttributeListRxt})?\*(\S|\S#{CC_ALL}*?\S)\*(?!#{CG_WORD})/m],
      # "`double-quoted`"
      [:double, :constrained, /(^|[^#{CC_WORD};:}])(?:#{QuoteAttributeListRxt})?"`(\S|\S#{CC_ALL}*?\S)`"(?!#{CG_WORD})/m],
      # '`single-quoted`'
      [:single, :constrained, /(^|[^#{CC_WORD};:`}])(?:#{QuoteAttributeListRxt})?'`(\S|\S#{CC_ALL}*?\S)`'(?!#{CG_WORD})/m],
      # ``monospaced``
      [:monospaced, :unconstrained, /\\?(?:#{QuoteAttributeListRxt})?``(#{CC_ALL}+?)``/m],
      # `monospaced`
      [:monospaced, :constrained, /(^|[^#{CC_WORD};:"'`}])(?:#{QuoteAttributeListRxt})?`(\S|\S#{CC_ALL}*?\S)`(?![#{CC_WORD}"'`])/m],
      # __emphasis__
      [:emphasis, :unconstrained, /\\?(?:#{QuoteAttributeListRxt})?__(#{CC_ALL}+?)__/m],
      # _emphasis_
      [:emphasis, :constrained, /(^|[^#{CC_WORD};:}])(?:#{QuoteAttributeListRxt})?_(\S|\S#{CC_ALL}*?\S)_(?!#{CG_WORD})/m],
      # ##mark## (referred to in AsciiDoc.py as unquoted)
      [:mark, :unconstrained, /\\?(?:#{QuoteAttributeListRxt})?##(#{CC_ALL}+?)##/m],
      # #mark# (referred to in AsciiDoc.py as unquoted)
      [:mark, :constrained, /(^|[^#{CC_WORD}&;:}])(?:#{QuoteAttributeListRxt})?#(\S|\S#{CC_ALL}*?\S)#(?!#{CG_WORD})/m],
      # ^superscript^
      [:superscript, :unconstrained, /\\?(?:#{QuoteAttributeListRxt})?\^(\S+?)\^/],
      # ~subscript~
      [:subscript, :unconstrained, /\\?(?:#{QuoteAttributeListRxt})?~(\S+?)~/],
    ]

    accum[true] = compat = normal.drop 0
    # ``quoted''
    compat[2] = [:double, :constrained, /(^|[^#{CC_WORD};:}])(?:#{QuoteAttributeListRxt})?``(\S|\S#{CC_ALL}*?\S)''(?!#{CG_WORD})/m]
    # `quoted'
    compat[3] = [:single, :constrained, /(^|[^#{CC_WORD};:}])(?:#{QuoteAttributeListRxt})?`(\S|\S#{CC_ALL}*?\S)'(?!#{CG_WORD})/m]
    # ++monospaced++
    compat[4] = [:monospaced, :unconstrained, /\\?(?:#{QuoteAttributeListRxt})?\+\+(#{CC_ALL}+?)\+\+/m]
    # +monospaced+
    compat[5] = [:monospaced, :constrained, /(^|[^#{CC_WORD};:}])(?:#{QuoteAttributeListRxt})?\+(\S|\S#{CC_ALL}*?\S)\+(?!#{CG_WORD})/m]
    # #unquoted#
    #compat[8] = [:unquoted, *compat[8][1..-1]]
    # ##unquoted##
    #compat[9] = [:unquoted, *compat[9][1..-1]]
    # 'emphasis'
    compat.insert 3, [:emphasis, :constrained, /(^|[^#{CC_WORD};:}])(?:#{QuoteAttributeListRxt})?'(\S|\S#{CC_ALL}*?\S)'(?!#{CG_WORD})/m]
  end

  # NOTE order of replacements is significant
  REPLACEMENTS = [
    # (C)
    [/\\?\(C\)/, '&#169;', :none],
    # (R)
    [/\\?\(R\)/, '&#174;', :none],
    # (TM)
    [/\\?\(TM\)/, '&#8482;', :none],
    # foo -- bar (where either space character can be a newline)
    # NOTE this necessarily drops the newline if replacement appears at end of line
    [/(?: |\n|^|\\)--(?: |\n|$)/, '&#8201;&#8212;&#8201;', :none],
    # foo--bar
    [/(#{CG_WORD})\\?--(?=#{CG_WORD})/, '&#8212;&#8203;', :leading],
    # ellipsis
    [/\\?\.\.\./, '&#8230;&#8203;', :none],
    # right single quote
    [/\\?`'/, '&#8217;', :none],
    # apostrophe (inside a word)
    [/(#{CG_ALNUM})\\?'(?=#{CG_ALPHA})/, '&#8217;', :leading],
    # right arrow ->
    [/\\?-&gt;/, '&#8594;', :none],
    # right double arrow =>
    [/\\?=&gt;/, '&#8658;', :none],
    # left arrow <-
    [/\\?&lt;-/, '&#8592;', :none],
    # left double arrow <=
    [/\\?&lt;=/, '&#8656;', :none],
    # restore entities
    [/\\?(&)amp;((?:[a-zA-Z][a-zA-Z]+\d{0,2}|#\d\d\d{0,4}|#x[\da-fA-F][\da-fA-F][\da-fA-F]{0,3});)/, '', :bounding],
  ]

  # Internal: Automatically load the Asciidoctor::Extensions module.
  #
  # Requires the Asciidoctor::Extensions module if the name is :Extensions.
  # Otherwise, delegates to the super method.
  #
  # This method provides the same functionality as using autoload on
  # Asciidoctor::Extensions, except that the constant isn't recognized as
  # defined prior to it being loaded.
  #
  # Returns the resolved constant, if resolved, otherwise nothing.
  def self.const_missing name
    if name == :Extensions
      require_relative 'asciidoctor/extensions'
      Extensions
    else
      super
    end
  end unless RUBY_ENGINE == 'opal'

  unless RUBY_ENGINE == 'opal'
    autoload :SyntaxHighlighter, %(#{__dir__}/asciidoctor/syntax_highlighter)
    autoload :Timings, %(#{__dir__}/asciidoctor/timings)
  end
end

# core extensions
require_relative 'asciidoctor/core_ext'

# modules and helpers
require_relative 'asciidoctor/helpers'
require_relative 'asciidoctor/logging'
require_relative 'asciidoctor/rx'
require_relative 'asciidoctor/substitutors'
require_relative 'asciidoctor/version'

# abstract classes
require_relative 'asciidoctor/abstract_node'
require_relative 'asciidoctor/abstract_block'

# concrete classes
require_relative 'asciidoctor/attribute_list'
require_relative 'asciidoctor/block'
require_relative 'asciidoctor/callouts'
require_relative 'asciidoctor/converter'
require_relative 'asciidoctor/document'
require_relative 'asciidoctor/inline'
require_relative 'asciidoctor/list'
require_relative 'asciidoctor/parser'
require_relative 'asciidoctor/path_resolver'
require_relative 'asciidoctor/reader'
require_relative 'asciidoctor/section'
require_relative 'asciidoctor/stylesheets'
require_relative 'asciidoctor/table'
require_relative 'asciidoctor/writer'

# main API entry points
require_relative 'asciidoctor/load'
require_relative 'asciidoctor/convert'

if RUBY_ENGINE == 'opal'
  require_relative 'asciidoctor/syntax_highlighter'
  require_relative 'asciidoctor/timings'
  # this require is satisfied by the Asciidoctor.js build; it supplies compile and runtime overrides for Asciidoctor.js
  require 'asciidoctor/js/postscript'
end
